attribution/attribution_metrics.py

```diff
@@ -0,0 +1,171 @@
+import math
+from typing import List, Tuple
+
+import numpy as np
+import openai
+from sklearn.metrics.pairwise import cosine_similarity
+from transformers import PreTrainedModel, PreTrainedTokenizer
+
+
+def token_prob_difference(
```

Comment:
I think this will work much better with nearest perturbation, since currently it collapses to near 0 and near 1.

We could also neatly modify this to look at the probability diff regardless of location in the output, right? For each token in the initial unperturbed output, if that token is in the perturbed output or *any* of their top 20 log probs for *any* perturbed output token, use that value. Does this make sense? That way it doesn't matter at all if the output length or position of the output token changes. (Though I can imagine this getting a bit tricky if tokens are repeated in the output).

Reply:
Yeah I thought about that, but repeated tokens may make this tricky. Common tokens will probably be repeated in the top 20 logprobs across the sentence but for unrelated reasons, artificially pushing the similarity score. 

We could also define a "window length" where to look for tokens, but I can imagine ways you can rephrase a sentence to have a very similar meaning but by shifting a token which were at the beginning of the sentence to the end. Looking at "nearby" tokens instead of the "same position" tokens wouldn't be robust in these cases.

Reply:
Mmm. It's tricky. Let's leave it for now - I'll put a ticket in the backlog and maybe we can figure out an elegant way to handle it later.

Reply:
Okay, sounds good!